### ISSUE TITLE
[obj-c] nil object error

### ISSUE THREAD

**Issue by nevir** [Microsoft/react-native-code-push#376]
I haven't been able to reproduce this one, but I just got a crash on a launch after a failed update (in the iOS simulator, code push 1.12.2, RN 0.27.1):

```
Jun 14 18:18:03 NevBook-2.local Driver[84012]: Update did not finish loading the last time, rolling back to a previous version.
Jun 14 18:18:03 NevBook-2.local Driver[84012]: *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSArrayM insertObject:atIndex:]: object cannot be nil'
	*** First throw call stack:
	(
		0   CoreFoundation                      0x00d3f946 __exceptionPreprocess + 182
		1   libobjc.A.dylib                     0x009caa97 objc_exception_throw + 44
		2   CoreFoundation                      0x00bf6c01 -[__NSArrayM insertObject:atIndex:] + 881
		3   CoreFoundation                      0x00bf6861 -[__NSArrayM addObject:] + 65
		4   CodePush                            0x0012469b -[CodePush saveFailedUpdate:] + 283
		5   CodePush                            0x001244fb -[CodePush rollbackPackage] + 139
		6   CodePush                            0x00123a55 -[CodePush initializeUpdateAfterRestart] + 341
		7   CodePush                            0x001238cd -[CodePush init] + 125
		8   libobjc.A.dylib                     0x009e097f +[NSObject new] + 98
		9   React                               0x003698c1 -[RCTModuleData setUpInstanceAndBridge] + 993
		10  React                               0x0036adec __25-[RCTModuleData instance]_block_invoke + 44
		11  React                               0x003fe7e4 RCTExecuteOnMainThread + 100
		12  React                               0x0036aba1 -[RCTModuleData instance] + 657
		13  React                               0x002f455b __49-[RCTBatchedBridge initModulesWithDispatchGroup:]_block_invoke_2 + 139
		14  libdispatch.dylib                   0x023aca2a _dispatch_call_block_and_release + 15
		15  libdispatch.dylib                   0x023c903f _dispatch_client_callout + 14
		16  libdispatch.dylib                   0x023b2764 _dispatch_main_queue_callback_4CF + 470
		17  CoreFoundation                      0x00c9995e __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 14
		18  CoreFoundation                      0x00c58760 __CFRunLoopRun + 2256
		19  CoreFoundation                      0x00c57bcb CFRunLoopRunSpecific + 443
		20  CoreFoundation                      0x00c579fb CFRunLoopRunInMode + 123
		21  GraphicsServices                    0x0588524f GSEventRunModal + 192
		22  GraphicsServices                    0x0588508c GSEventRun + 104
		23  UIKit                               0x010f18b6 UIApplicationMain + 1526
		24  Driver                              0x000abf2a main + 138
		25  libdyld.dylib                       0x023eeac9 start + 1
```

Appears to be at https://github.com/Microsoft/react-native-code-push/blob/master/ios/CodePush/CodePush.m#L433

**Comment by lostintangent** [comment]
Interesting. I can see from the stack why this exception could occur, but I'm not sure how to get into a state where it would be possible.

Was this a release or debug build? Was your AppDelegate.m configured to get the JS bundle location from CodePush or from the packager?

**Comment by nevir** [comment]
Production build; bundle uri came from code push.

Still can't repro :(

**Comment by lostintangent** [comment]
Any chance you can recall the sequence of steps you took that led to this issue? Did you deploy a production build to the simulator, then install a CodePush update and then hit this issue? Did you re-deploy a binary update between the time you installed the update and restarted the app to apply it?

**Comment by nevir** [comment]
I believe the steps I took were:

1. deploy to simulator w/ baked in bundle
2. new bundle was available; it was downloaded and applied (immediately)
3. the new bundle was missing a call to `notifyAppReady`
4. re-deployed a new copy to the simulator
5. app crashed on first launch

**Comment by lostintangent** [comment]
@nevir Have you seen this issue occur again? I believe it is actually related to #378, and therefore, if we fix the root cause there, where calling `install` twice on the same update can get an app into a bad state, then it should prevent this from happening as well.

Without a reliable repro (which we haven't been able to get), we can't say for sure, but I'm fairly confident that the double `install` bug leads to general state corruption of the CodePush plugin, and preventing it should prevent numerous resulting bugs such as this.

**Comment by nevir** [comment]
Haven't seen it repeat

**Comment by lostintangent** [comment]
OK cool. I'm going to close this as resolved via #407 then, but please re-open this if you see it happen. Thanks again!

**Comment by nevir** [comment]
👍
